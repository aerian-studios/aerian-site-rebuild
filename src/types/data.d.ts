import * as React from "react";

export const isImageSharp = (
    image: ImageSharp | string
): image is ImageSharp => {
    return typeof image !== "string";
};

export interface Image {
    alt?: string;
    image: string | ImageSharp;
}

export interface ImageSharp {
    childImageSharp: {
        sizes: ImageSharpSizes;
        src: string;
    };
}

export interface ImageSharpSizes {
    aspectRatio?: number;
    sizes?: string;
    src: string;
    srcSet?: string;
    tracedSVG?: string;
}
