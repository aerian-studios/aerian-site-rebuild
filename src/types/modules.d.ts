declare module "*.svg";
declare module "*.png";
declare module "*.json";

// For storyshots
declare module "babel-plugin-require-context-hook/register";

// for maps
declare module "pigeon-overlay";
declare module "pigeon-maps";
