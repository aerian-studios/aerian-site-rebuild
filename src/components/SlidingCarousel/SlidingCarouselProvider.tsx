import classNames from "classnames";
import debounce from "debounce";
import * as React from "react";
import uuid from "uuid";

import { SliderProps, SlidingCarousel } from "./SlidingCarousel";

import * as styles from "./SlidingCarousel.module.scss";

interface ButtonProps extends React.HTMLAttributes<HTMLButtonElement> {}
export type ButtonRenderProp<P extends ButtonProps = ButtonProps> = (
    button: React.JSXElementConstructor<P>
) => React.ReactElement;

interface Props extends SliderProps {
    carouselLabel: string | React.ReactElement;
    buttonBackContentRender?: ButtonRenderProp;
    buttonFwdContentRender?: ButtonRenderProp;
    center?: boolean;
    infinite?: boolean;
    itemGap?: string;
    className?: string;
    style?: React.CSSProperties;
}

interface Size {
    width: number;
    height: number;
}

/**
 * Utility for setting CSS variables on the root node
 * @param {String} propertyName - String name of the CSS variable - e.g. --SCCardHeight
 * @param {String} propertyValue - value to set
 * @returns {void}
 */
const setCSSVariable = (propertyName: string, propertyValue: string): void => {
    const body = document.body;

    if (!body) {
        return;
    }

    body.style.setProperty(propertyName, propertyValue);
};

const workOutSizes = (nodes: HTMLCollection): Size[] => {
    const len = nodes.length;
    const ret: Size[] = [];

    if (len) {
        for (let i = 0; i < len; i++) {
            const el: Element = nodes[i];

            ret.push({ width: el.clientWidth, height: el.clientHeight });
        }
    }
    return ret;
};

export const SlidingCarouselProvider: React.FC<Props> = ({
    carouselLabel,
    children,
    className,
    style,
    buttonBackContentRender,
    buttonFwdContentRender,
    center = true,
    infinite = false,
    itemGap = "0px"
}) => {
    const sliderRef = React.useRef<HTMLDivElement>();
    let stillInFrame = false; // used for some render optimisations
    let originalWidth = 0; // used largely in infinite scroll
    let originalChildLength = 0; // used largely in infinite scroll
    let pageWidth = 200;
    let childSizes: Size[] = []; // a utitlity array for more speedy checking of child sizes
    let current = 0; // the current index
    let destination = 0; // destination in async functions

    const setCurrent = (newCurrent: number): number => {
        current = newCurrent !== current ? newCurrent : current;
        setCurrentFocus(current);

        return current;
    };

    const setCurrentFocus = (index: number): void => {
        if (!sliderRef || !sliderRef.current) {
            return;
        }
        const childs = sliderRef.current.children;

        const focussable = childs[current].querySelector("a");
        // tslint:disable-next-line
        focussable && focussable.focus();
    };

    interface LoopCheck {
        position: number | null;
    }
    /**
     * Method to check if the infinite scroll needs to loop
     * @param {number} position - the current scroll position
     * @return {LoopCheck} - A position of either `null` for no loop need or an adjusted position number that doesn't need to loop
     */
    const checkForLoop = (position: number): LoopCheck => {
        const loopCheck: LoopCheck = { position: null };

        if (position < originalWidth) {
            loopCheck.position = position + originalWidth;
        } else if (position > originalWidth * 2) {
            loopCheck.position = position - originalWidth;
        }

        return loopCheck;
    };

    const alignThingsNearestToPosition = (position?: number) => {
        if (!childSizes.length || position === destination) {
            return;
        }

        destination = typeof position !== "undefined" ? position : destination;
        let scrollAmount = center ? -pageWidth * 0.5 : 0;
        let i = 0;
        let elW = 0;
        const gap = parseInt(itemGap, 10);

        // assume that the children have non-uniform widths
        while (scrollAmount < destination) {
            elW = childSizes[i].width + gap;
            i++;
            if (i === childSizes.length - 1) {
                break;
            }

            scrollAmount += elW;
        }

        if (center) {
            scrollAmount += elW * 0.5 + gap / 2;
        }

        if (i === current) {
            return;
        }
        destination = scrollAmount;
        setCurrent(i);
        scrollToPosition(scrollAmount);
    };

    const debouncedCentreThingsNearestToPosition = debounce(
        alignThingsNearestToPosition,
        400
    );

    const alignTheThingsToIndex = (indexToCentre?: number): void => {
        if (!childSizes.length) {
            return;
        }

        const index = indexToCentre ? indexToCentre : current;
        let scrollAmount = center ? -pageWidth * 0.5 : 0;
        let elW = 0;
        const gap = parseInt(itemGap, 10);

        for (let i = 0; i < index; i++) {
            elW = childSizes[i].width + gap;

            scrollAmount += elW;
        }

        if (center) {
            scrollAmount += elW * 0.5 + gap / 2;
        }

        // tslint:disable-next-line
        if (index !== current) {
            destination = scrollAmount;
            scrollToPosition(destination);
            setCurrent(index);
        }
    };

    const scrollToPosition = (
        position: number,
        immediate: boolean = false
    ): void => {
        if (!sliderRef || !sliderRef.current) {
            return;
        }

        if (immediate) {
            sliderRef.current.style.scrollBehavior = "auto";
        }
        sliderRef.current.scroll(position, 0);
        if (immediate) {
            sliderRef.current.style.scrollBehavior = "smooth";
        }
    };

    const alignTheThings = () => {
        pageWidth = window.innerWidth;
        alignTheThingsToIndex(current);
    };

    let deferDelay: any;
    let scrollDelay: any;
    const moveCarouselPage = (direction: number) => {
        // tslint:disable-next-line
        deferDelay && window.clearTimeout(deferDelay);
        // tslint:disable-next-line
        scrollDelay && window.clearTimeout(scrollDelay);

        const visibleItems = childSizes.slice(current - 1, current + 1);
        const gap = parseInt(itemGap, 10);
        let amountToScroll = 0;
        let w = (pageWidth - visibleItems[1].width) / 2;

        while (visibleItems.length && amountToScroll <= w) {
            const itemIndex =
                visibleItems.length > 1
                    ? Math.ceil(visibleItems.length / 2)
                    : 0;

            const item = visibleItems.splice(itemIndex, 1);

            // update page centre to item centre based on item width
            w = (pageWidth - item[0].width) / 2;
            amountToScroll += item[0].width / 2 + gap / 2;
        }

        // Check for loop before scrolling
        destination = destination + amountToScroll * direction;
        const destCheck = checkForLoop(destination);

        // disable the scroll centerer
        stillInFrame = true;

        if (destCheck.position) {
            destination = destCheck.position;
            const preDest = destination - amountToScroll * direction;
            scrollToPosition(preDest, true);
        }

        // Needs to defer this second scroll request
        deferDelay = window.setTimeout(() => {
            scrollToPosition(destination);
        }, 60);

        // wait until scroll complete before re-enabling scroll checking
        scrollDelay = window.setTimeout(() => {
            stillInFrame = false;
        }, 2000);
    };

    /**
     * Handle scroll events by calling necessary checks
     * @return {void}
     */
    const scrollHandler = (): void => {
        if (!sliderRef || !sliderRef.current) {
            return;
        }
        const lastX = sliderRef.current.scrollLeft;

        if (!stillInFrame && infinite) {
            window.requestAnimationFrame(() => {
                const check = checkForLoop(lastX);

                if (check.position) {
                    scrollToPosition(check.position, true);
                }
                stillInFrame = false;
            });

            stillInFrame = true;
        }

        debouncedCentreThingsNearestToPosition(lastX);
    };

    // HOOKS
    /**
     * Set some init code to workout original width and do any centring that needs to happen
     */
    React.useLayoutEffect(() => {
        if (!sliderRef || !sliderRef.current) {
            return;
        }
        originalWidth = sliderRef.current.scrollWidth / 3;
        pageWidth = window.innerWidth;

        childSizes = workOutSizes(sliderRef.current.children);
        originalChildLength = infinite
            ? childSizes.length / 3
            : childSizes.length;

        // This also needs to be deferred :(|)
        window.setTimeout(() => {
            alignTheThingsToIndex(
                center ? Math.floor((childSizes.length - 1) / 2) : 0
            );
        }, 500);
    });

    /**
     * Add scroll cheking on slider
     */
    React.useEffect(() => {
        if (!sliderRef || !sliderRef.current) {
            return;
        }
        // set scroll position to the middle
        sliderRef.current.scroll(originalWidth, 0);
        sliderRef.current.addEventListener("scroll", scrollHandler);

        // Remove the event listener on destroy
        return () => {
            // tslint:disable-next-line
            sliderRef.current &&
                sliderRef.current.removeEventListener("scroll", scrollHandler);
            debouncedCentreThingsNearestToPosition.clear();
        };
    }, [sliderRef.current]);

    React.useEffect(() => {
        window.addEventListener("resize", alignTheThings);

        return () => {
            window.removeEventListener("resize", alignTheThings);
        };
    });

    /**
     * Set some CSS variables based on props
     */
    React.useEffect(() => {
        setCSSVariable("--SCSnapAlign", center ? "center" : "start");
        setCSSVariable("--SCGridGap", itemGap);
    });

    const FwdButton: React.SFC<ButtonProps> = props => (
        <button
            aria-hidden="true"
            tabIndex={-1}
            {...props}
            onClick={() => {
                if (!sliderRef || !sliderRef.current) {
                    return;
                }

                moveCarouselPage(1);
            }}
        />
    );
    const BackButton: React.SFC<ButtonProps> = props => (
        <button
            aria-hidden="true"
            tabIndex={-1}
            {...props}
            onClick={() => {
                if (!sliderRef || !sliderRef.current) {
                    return;
                }

                moveCarouselPage(-1);
            }}
        />
    );

    interface SliderProgressTrackerProps
        extends React.AllHTMLAttributes<HTMLElement> {
        index: number;
        total: number;
        itemTitles?: string[];
        PagerElement?: React.JSXElementConstructor<{
            current: boolean;
            children: JSX.Element;
        }>;
        pagerElementClick: (index: number) => void;
    }
    const SliderProgressTracker: React.SFC<SliderProgressTrackerProps> = ({
        index,
        total,
        itemTitles,
        PagerElement,
        pagerElementClick
    }) => {
        return (
            <div>
                <p
                    className={styles.SCProgressFraction}
                    aria-live="polite"
                    aria-atomic="true"
                >{`Item ${index} of ${total}`}</p>
                {PagerElement && itemTitles ? (
                    <ol>
                        {itemTitles.map((item, i) => (
                            <li
                                className={classNames(styles.SCPagerElement)}
                                key={item}
                            >
                                <PagerElement current={i === current}>
                                    <button
                                        onClick={() => {
                                            pagerElementClick(i);
                                        }}
                                    >
                                        <span className="visuallyhidden">
                                            {item}
                                        </span>{" "}
                                        {i !== current ? null : (
                                            <span className="visuallyhidden">
                                                (Current item)
                                            </span>
                                        )}
                                    </button>
                                </PagerElement>
                            </li>
                        ))}
                    </ol>
                ) : null}
            </div>
        );
    };

    const fwdButton = buttonFwdContentRender
        ? buttonFwdContentRender(FwdButton)
        : null;
    const backButton = buttonBackContentRender
        ? buttonBackContentRender(BackButton)
        : null;
    const uniqueId = uuid();
    const regionLabel =
        typeof carouselLabel === "string"
            ? { "aria-label": carouselLabel }
            : { "aria-labelledby": uniqueId };

    return (
        <section
            className={classNames(className, styles.sliderProvider)}
            style={style}
            {...regionLabel}
        >
            {typeof carouselLabel === "string" ? null : carouselLabel}
            <SlidingCarousel infinite={infinite} ref={sliderRef}>
                {children}
            </SlidingCarousel>
            {backButton}
            {fwdButton}
            <SliderProgressTracker
                index={current}
                total={originalChildLength}
                pagerElementClick={alignTheThingsToIndex}
            />
        </section>
    );
};
