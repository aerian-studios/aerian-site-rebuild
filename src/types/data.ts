import { Stats } from "fs";
import { ParsedPath } from "path";

export type ImageField = string | ImageSharp;
export type FileField = string | Partial<FileNode>;

// Images

export interface GalleryImage {
    alt?: string;
    image: ImageField;
}

export interface FileNode extends ParsedPath, Stats {
    publicURL: string;
    id: string;
    children: any[];
    sourceInstanceName: string;
    absolutePath: string;
    relativePath: string;
    relativeDirectory: string;
    extension: string;
    size: number;
    prettySize: string;

    modifiedTime: string;
    accessTime: string;
    changeTime: string;
    birthTime: string;
    internal: {
        contentDigest: string;
        type: "Directory" | "File";
        mediaType?: string;
        description: string;
    };
}

export interface ImageSharp extends Partial<FileNode> {
    childImageSharp: {
        fluid?: ImageSharpFluid;
        fixed?: ImageSharpFixed;
    };
}
export interface ImageSharpFluid {
    aspectRatio?: number;
    sizes?: string;
    src: string;
    srcSet?: string;
    tracedSVG?: string;
    base64?: string;
}

export interface ImageSharpFixed {
    aspectRatio?: number;
    src: string;
    srcSet?: string;
    tracedSVG?: string;
    base64?: string;
    width: number;
    height: number;
}

// PAGES

export interface NodeList<T> {
    edges: Array<Edge<T>>;
}

export interface Edge<T> {
    node: T;
}
export interface PagesListData {
    data: {
        allPagesJson: PageList;
    };
}

export interface PageList {
    edges: PageListNode[];
}

export type PageListNode = Edge<PageRef>;

export interface PageRef {
    id: string;
    path: string;
    title: string;
}

export interface PageSection {
    title: string;
    image?: ImageField;
    smallImage?: ImageField;
    subtitle: string;
    blurb: string;
    activities?: Activity[];
    successStory?: SuccessStory;
    testimonial?: Testimonial;
}

export interface Page {
    title: string;
    path: string;
    heroImage?: ImageField;
    subheading: string;
    description: string;
    seoTitle: string;
    seoDescription: string;
    seoKeywords: string;
}

export interface WhatWeDo extends Page {
    sections?: PageSection[];
}

export interface MeetTheTeam extends Page {
    staff: Staff[];
}

export interface About extends Page {
    infographic?: Infographic[];
}

export type InfographicType =
    | "Single Line"
    | "Compact Single Line"
    | "Split"
    | "Bar"
    | "Vertical"
    | "Horizontal"
    | "Surround";
export interface Infographic {
    primaryText?: string;
    type: InfographicType;
    secondaryText?: string;
    image?: ImageField;
    imageCount: number;
    secondaryImage?: ImageField;
}

export interface Staff {
    name: string;
    jobTitle: string;
    live: boolean;
    imageNormal: ImageField;
    imageFunny: ImageField;
    description: string;
    fact: string;
    skills: string[];
}

export interface ReactRouterLocation {
    action: string;
    hash: string;
    key: string;
    pathname: string;
    search: string;
    state: null | string;
}

export interface Testimonial {
    quote?: string;
    person?: string;
    title?: string;
    avatar?: ImageField;
    image?: ImageField;
}

export interface Stat {
    title: string;
    text: string;
    image: ImageField;
}
export interface SuccessStory {
    title: string;
    stats?: Stat[];
}

export interface Activity {
    title: string;
    text: string;
}

export interface Hurdle {
    title: string;
    image: ImageField;
    text: string;
}

export interface ProjectStage {
    description: string;
    hurdles: Hurdle[];
}

export interface Performance {
    title: string;
    text: string;
}
export interface ProjectBox {
    name: string;
    slug: string;
    heroImage: ImageField;
    thumbnail: ImageField;
    client: Client;
    titleLineOne: string;
}
export interface Project {
    name: string;
    slug: string;
    heroImage: ImageField;
    heroVideo?: string;
    titleLineOne: string;
    titleLineTwo?: string;
    client: Client;
    goLiveDate?: string;
    caseStudyTitle: string;
    caseStudyText: string;
    caseStudyImage: ImageField;
    externalUrl: string;
    rolloverDetails: string;
    homepage: boolean;
    featured: boolean;
    gallery: GalleryImage[];
    challenge: ProjectStage;
    solution: ProjectStage;
    results: ProjectStage;
    performance?: Performance[];
    testimonial?: Testimonial;
}

export interface Client {
    name: string;
    logo: ImageField;
    promoLogo: ImageField;
    featured: boolean;
    featuredProject?: string;
    id: string;
}
