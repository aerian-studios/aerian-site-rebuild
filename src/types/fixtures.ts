import {
    About,
    Client,
    ImageField,
    ImageSharp,
    Infographic,
    MeetTheTeam,
    PageSection,
    Performance,
    Project,
    ProjectBox,
    SuccessStory,
    Testimonial,
    WhatWeDo
} from "./data";

import * as aboutData from "../pages/about.json";
import * as clientData from "../pages/clients/cnn.json";
import * as meetTheTeamData from "../pages/meet-the-team.json";
import * as projectData from "../pages/projects/bbc-iwonder-perfect-diet-for-you.json";
import * as whatWeDoData from "../pages/what-we-do.json";

import { absolutifyURLs } from "../lib/helpers";
export const client = absolutifyURLs<Client>(clientData);

export const project = { ...absolutifyURLs<Project>(projectData), client };
export const projectBox: ProjectBox = {
    name: project.name,
    slug: project.slug,
    titleLineOne: project.titleLineOne,
    thumbnail: project.heroImage,
    heroImage: project.heroImage,
    client
};
export const whatWeDo = absolutifyURLs<Required<WhatWeDo>>(whatWeDoData);
export const about = absolutifyURLs<About>(aboutData);
export const meetTheTeam = absolutifyURLs<MeetTheTeam>(meetTheTeamData);
export const projectStage = project.challenge;
export const hurdles = projectStage.hurdles;
export const hurdle = hurdles[0];
export const performances: Performance[] = project.performance as Performance[];
export const performance = performances[0];
export const testimonial: Testimonial = project.testimonial as Testimonial;
export const testimonialSection: PageSection = whatWeDo.sections[0];
export const successStorySection: PageSection = whatWeDo.sections[1];
export const successStory: SuccessStory = successStorySection.successStory as SuccessStory;
export const pageSection = {
    ...successStorySection,
    testimonial: testimonialSection.testimonial
} as Required<PageSection>;
export const activity = pageSection.activities[0];
export const stat = successStory.stats![0];
export const staff = meetTheTeam.staff;
export const person = staff[0];
export const infographics = about.infographic as Infographic[];
export const infographic = infographics[0];

export const image: string = person.imageFunny as string;
export const imageSharpFluid: ImageSharp = {
    childImageSharp: {
        fluid: {
            aspectRatio: 3 / 4,
            sizes: "(max-width: 290px) 100vw, 290px",
            base64:
                "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAYCAYAAAD6S912AAAACXBIWXMAAAsSAAALEgHS3X78AAAEe0lEQVQ4y42Te1BUVRzH70z/9V9ZToXy2GXfD3Z57C4sDyHtreWMIqmEgUVCak4qwiiRiausiIhk6JLiAzXGEE1FBVMaEBtsrJnQwsEaNTMbmxRrFJZP516EWXykZ+Y3d/fc7/mc7+9xJQJWf3+/8vT7/dy9BvdkzaDufkt6EKznxnXKS0uo+rSCP69eVfb6+vqGaf8XGCjs6enhtZdeRJIkHn9MwmGzcv589yNBhzns7e1VfnuKP2GcTY8nO51Vs96iaNp4ctImcvn3K0NZPBJQXn9fv8EHGWkcKfuIrrpqftmzlXO1VVTNnUG1SD/Q5QOBMszvHxB9297G4sxUjlV5+bqylK76Lzi2bjUNxflsWeOhr1858HCH/ju3fnP8KMU502mqWMmmRR/Sss7L/pKl1OTPY/daDzf/+ffhNRxwONDd002H2LDofVrWe2kuW8FRz8ecKPPQXuGhtVwAL10cAN5ntIYBBwW/7d9Hw7ICftjpo2tLJY1FBTQVL6FzUyXt4oKe7u6HAwMFt8SBI6XLOSAcndpcSUt5CQ3zFnCgpJDmSi83/7r2aCkHujwrGrPLs4RD65bTurGcVpHyHs9CTjTsHLr8QV/M8MG+A7z2xxWqlxbgK5zHcd9qmtcuY33eLH463aG87+29PQS7+ykF3jRws+j47Vs0VJRQNieLls1llOZm8NnC9zjbcXKYG3ke5Qgc9CHg4At5nTzZQXVRPnVF8+ms89HiK6XN52V1QR4rSsro7Dxzz3APQiX5RWAtjhxuIjnlZTInTaamYDY/1lXx84FaOr+spjAnmxFPP4fZFMHENyZTXLySxoOHuXDh4r01bGtrJzPzXWKi43A43FgskazJn0tn/Qa6m3fzff1WslJT0WgNGAwmgoPDGD0qjHC1AacjnqKiZcK1H2n79h1kZ+cSGelArzPhcsUTF5vAE089S17GJM41bKTr8C6+27uT1PGvowrXodHIoUevN2G1RmI22xUjly5dRgoOVhEaqhab1jsRgcVqEw5C8eXN5PzeDZxp3MHBmhqy3pxKiNBqwrWo1Vq0wq3ZbBPQKGEiidraXUgGgxk5jEazkoocJgFOco/h88X5/LpvG6f2bmP/1t3MnpZOmFqDTmdU3JlMVgVoMgmoJUqUKh5pVFCIIpChOhFarR6bLYqEhOdxu8eSnTaDqhVeWr86Rm5WNuEiXaPJImpsE7oY7CJkt0aDlVhXItKIJ0diMloYHaLmnSmT8M6fI+oRS0J8ihIOIYpzp7AwdwFZU6djEFr5Qrs9huiYWHE2QpTAIHrgxBHjRhopxkCl1vNCUhKN61fx9pQ0YoT1ROEw3p3MmKRxxMYmKk6S3fHY7NFKA+WQG6JSaZS0o6JcRAsj0jMjgwgarSLt1VdYPDODUDEGcoFdrgSRdgqJiWPR6S2kTUwlfcIELBGRwkmccOhQQGGh4ei0RsVhpN2JFBoSTliYFrslgmSnUznsFHPociYoMLkuKpWWKJGiS4BstmglXbmzBoOFoKBgAdUQIf7L8R+bum8HTMuX7wAAAABJRU5ErkJggg==",
            src: image,
            srcSet: image
        }
    }
};

export const imageSharpFixed: ImageSharp = {
    childImageSharp: {
        fixed: {
            width: 290,
            height: 290,
            base64:
                "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAYCAYAAAD6S912AAAACXBIWXMAAAsSAAALEgHS3X78AAAEe0lEQVQ4y42Te1BUVRzH70z/9V9ZToXy2GXfD3Z57C4sDyHtreWMIqmEgUVCak4qwiiRiausiIhk6JLiAzXGEE1FBVMaEBtsrJnQwsEaNTMbmxRrFJZP516EWXykZ+Y3d/fc7/mc7+9xJQJWf3+/8vT7/dy9BvdkzaDufkt6EKznxnXKS0uo+rSCP69eVfb6+vqGaf8XGCjs6enhtZdeRJIkHn9MwmGzcv589yNBhzns7e1VfnuKP2GcTY8nO51Vs96iaNp4ctImcvn3K0NZPBJQXn9fv8EHGWkcKfuIrrpqftmzlXO1VVTNnUG1SD/Q5QOBMszvHxB9297G4sxUjlV5+bqylK76Lzi2bjUNxflsWeOhr1858HCH/ju3fnP8KMU502mqWMmmRR/Sss7L/pKl1OTPY/daDzf/+ffhNRxwONDd002H2LDofVrWe2kuW8FRz8ecKPPQXuGhtVwAL10cAN5ntIYBBwW/7d9Hw7ICftjpo2tLJY1FBTQVL6FzUyXt4oKe7u6HAwMFt8SBI6XLOSAcndpcSUt5CQ3zFnCgpJDmSi83/7r2aCkHujwrGrPLs4RD65bTurGcVpHyHs9CTjTsHLr8QV/M8MG+A7z2xxWqlxbgK5zHcd9qmtcuY33eLH463aG87+29PQS7+ykF3jRws+j47Vs0VJRQNieLls1llOZm8NnC9zjbcXKYG3ke5Qgc9CHg4At5nTzZQXVRPnVF8+ms89HiK6XN52V1QR4rSsro7Dxzz3APQiX5RWAtjhxuIjnlZTInTaamYDY/1lXx84FaOr+spjAnmxFPP4fZFMHENyZTXLySxoOHuXDh4r01bGtrJzPzXWKi43A43FgskazJn0tn/Qa6m3fzff1WslJT0WgNGAwmgoPDGD0qjHC1AacjnqKiZcK1H2n79h1kZ+cSGelArzPhcsUTF5vAE089S17GJM41bKTr8C6+27uT1PGvowrXodHIoUevN2G1RmI22xUjly5dRgoOVhEaqhab1jsRgcVqEw5C8eXN5PzeDZxp3MHBmhqy3pxKiNBqwrWo1Vq0wq3ZbBPQKGEiidraXUgGgxk5jEazkoocJgFOco/h88X5/LpvG6f2bmP/1t3MnpZOmFqDTmdU3JlMVgVoMgmoJUqUKh5pVFCIIpChOhFarR6bLYqEhOdxu8eSnTaDqhVeWr86Rm5WNuEiXaPJImpsE7oY7CJkt0aDlVhXItKIJ0diMloYHaLmnSmT8M6fI+oRS0J8ihIOIYpzp7AwdwFZU6djEFr5Qrs9huiYWHE2QpTAIHrgxBHjRhopxkCl1vNCUhKN61fx9pQ0YoT1ROEw3p3MmKRxxMYmKk6S3fHY7NFKA+WQG6JSaZS0o6JcRAsj0jMjgwgarSLt1VdYPDODUDEGcoFdrgSRdgqJiWPR6S2kTUwlfcIELBGRwkmccOhQQGGh4ei0RsVhpN2JFBoSTliYFrslgmSnUznsFHPociYoMLkuKpWWKJGiS4BstmglXbmzBoOFoKBgAdUQIf7L8R+bum8HTMuX7wAAAABJRU5ErkJggg==",
            src: image,
            srcSet: image
        }
    }
};
export const imageSharp = imageSharpFluid;
export const imageField: ImageField = imageSharp;
