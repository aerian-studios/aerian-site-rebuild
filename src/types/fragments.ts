import { graphql } from "gatsby";

export const pageFieldsFragment = graphql`
    fragment PageFields on PagesJson {
        title
        heroImage {
            childImageSharp {
                fluid(srcSetBreakpoints: [1024, 720, 360]) {
                    ...GatsbyImageSharpFluid_withWebp
                }
            }
        }
        subheading
        description
        seoTitle
        seoDescription
        seoKeywords
    }
`;

export const clientFragment = graphql`
    fragment Client on ClientsJson {
        name
        logo {
            childImageSharp {
                fluid(maxWidth: 600, srcSetBreakpoints: [420, 210]) {
                    ...GatsbyImageSharpFluid_withWebp_tracedSVG
                }
            }
            publicURL
        }
        promoLogo {
            childImageSharp {
                fluid(maxWidth: 600, srcSetBreakpoints: [420, 210]) {
                    ...GatsbyImageSharpFluid_withWebp_tracedSVG
                }
            }
            publicURL
        }
        featured
        featuredProject
    }
`;

export const projectBoxFragment = graphql`
    fragment ProjectBox on ProjectsJson {
        name
        slug
        titleLineOne
        client {
            ...Client
        }
        heroImage {
            childImageSharp {
                fluid(
                    maxWidth: 1066
                    maxHeight: 1066
                    srcSetBreakpoints: [1066, 720, 360, 250]
                ) {
                    ...GatsbyImageSharpFluid_withWebp_noBase64
                }
            }
            publicURL
        }
    }
`;

export const projectFragment = graphql`
    fragment Project on ProjectsJson {
        name
        slug
        client {
            ...Client
        }
        titleLineOne
        titleLineTwo
        goLiveDate
        caseStudyTitle
        caseStudyText
        externalUrl
        heroImage {
            childImageSharp {
                fluid(srcSetBreakpoints: [1024, 720, 360]) {
                    ...GatsbyImageSharpFluid_withWebp
                }
            }
            publicURL
        }
        heroVideo {
            publicURL
        }
        gallery {
            image {
                childImageSharp {
                    fluid(srcSetBreakpoints: [720, 360]) {
                        ...GatsbyImageSharpFluid_withWebp
                    }
                }
                publicURL
            }
            alt
        }
        challenge {
            description
            hurdles {
                title
                image {
                    childImageSharp {
                        fixed(width: 290) {
                            ...GatsbyImageSharpFixed_withWebp_tracedSVG
                        }
                    }
                    publicURL
                }
                text
            }
        }
        solution {
            hurdles {
                title
                image {
                    childImageSharp {
                        fixed(width: 290) {
                            ...GatsbyImageSharpFixed_withWebp_tracedSVG
                        }
                    }
                    publicURL
                }
                text
            }
            description
        }
        results {
            description
            hurdles {
                title
                image {
                    childImageSharp {
                        fixed(width: 290) {
                            ...GatsbyImageSharpFixed_withWebp_tracedSVG
                        }
                    }
                    publicURL
                }
                text
            }
        }
        performance {
            title
            text
        }
        testimonial {
            quote
            person
            title
            avatar {
                childImageSharp {
                    fixed(width: 75) {
                        ...GatsbyImageSharpFixed_withWebp_tracedSVG
                    }
                }
                publicURL
            }
        }
        id
    }
`;
