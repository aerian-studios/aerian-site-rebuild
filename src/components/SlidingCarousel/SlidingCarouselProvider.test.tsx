/// <reference types="@types/jest" />
import * as React from "react";

import * as renderer from "react-test-renderer";

import { projectBox } from "../../types/fixtures";
import {
    calculateNearestSnapPoint,
    calculateScrollOffsetForIndex,
    SlidingCarouselProvider
} from "./index";
import { Size } from "./SlidingCarouselProvider";

describe("SlidingCarouselProvider", () => {
    const childSizes: Size[] = [
        { width: 106, height: 306 },
        { width: 206, height: 306 },
        { width: 306, height: 306 },
        { width: 406, height: 306 }
    ];

    it("renders correctly", () => {
        const tree = renderer
            .create(<SlidingCarouselProvider carouselLabel="SUT" />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("renders correctly with infinite set", () => {
        const tree = renderer
            .create(
                <SlidingCarouselProvider infinite={true} carouselLabel="SUT" />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("correctly calculates index offset, left aligned", () => {
        const first = calculateScrollOffsetForIndex(0, childSizes, 0);
        const second = calculateScrollOffsetForIndex(1, childSizes, 0);
        const third = calculateScrollOffsetForIndex(2, childSizes, 0);
        const fourth = calculateScrollOffsetForIndex(3, childSizes, 0);

        expect(first).toEqual(0);
        expect(second).toEqual(106);
        expect(third).toEqual(312);
        expect(fourth).toEqual(618);
    });

    it("correctly calculates index offset, centre aligned", () => {
        const first = calculateScrollOffsetForIndex(0, childSizes, 300);
        const second = calculateScrollOffsetForIndex(1, childSizes, 300);
        const third = calculateScrollOffsetForIndex(2, childSizes, 300);
        const fourth = calculateScrollOffsetForIndex(3, childSizes, 300);
        expect(first).toEqual(-300 + childSizes[0].width / 2);
        expect(second).toEqual(
            -300 + childSizes[0].width + childSizes[1].width / 2
        );
        expect(third).toEqual(
            -300 +
                childSizes[0].width +
                childSizes[1].width +
                childSizes[2].width / 2
        );
        expect(fourth).toEqual(
            -300 +
                childSizes[0].width +
                childSizes[1].width +
                childSizes[2].width +
                childSizes[3].width / 2
        );
    });

    it("correctly calculates snap offset, left aligned", () => {
        const first = calculateNearestSnapPoint(0, childSizes, 0);
        const second = calculateNearestSnapPoint(300, childSizes, 0);
        const third = calculateNearestSnapPoint(309, childSizes, 0);
        const fourth = calculateNearestSnapPoint(600, childSizes, 0);

        expect(first).toEqual(0);
        expect(second).toEqual(312);
        expect(third).toEqual(312);
        expect(fourth).toEqual(618);
    });

    it("correctly calculates snap offset, centre aligned", () => {
        const first = calculateNearestSnapPoint(0, childSizes, 300);
        const second = calculateNearestSnapPoint(300, childSizes, 300);
        const third = calculateNearestSnapPoint(309, childSizes, 300);
        const fourth = calculateNearestSnapPoint(600, childSizes, 300);

        expect(first).toEqual(-300 + childSizes[0].width / 2);
        expect(second).toEqual(-300 + 312 + childSizes[2].width / 2);
        expect(third).toEqual(-300 + 312 + childSizes[2].width / 2);
        expect(fourth).toEqual(-300 + 618 + childSizes[3].width / 2);
    });
});
