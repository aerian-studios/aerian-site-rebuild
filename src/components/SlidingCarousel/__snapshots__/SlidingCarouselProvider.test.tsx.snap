// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SlidingCarouselProvider renders correctly 1`] = `
<section
  aria-label="SUT"
  className=""
>
  <div
    className=""
  >
    <div
      className=""
    />
  </div>
  <div>
    <p
      aria-atomic="true"
      aria-live="polite"
    >
      Item 0 of 0
    </p>
  </div>
</section>
`;

exports[`SlidingCarouselProvider renders correctly with infinite set 1`] = `
<section
  aria-label="SUT"
  className=""
>
  <div
    className=""
  >
    <div
      className=""
    />
  </div>
  <div>
    <p
      aria-atomic="true"
      aria-live="polite"
    >
      Item 0 of 0
    </p>
  </div>
</section>
`;
